src/tools/rbd_mirror/ImageDeleter.h

```diff
@@ -0,0 +1,122 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2016 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef CEPH_RBD_MIRROR_IMAGEDELETER_H
+#define CEPH_RBD_MIRROR_IMAGEDELETER_H
+
+#include <queue>
+#include "include/atomic.h"
+#include "common/Mutex.h"
+#include "common/Cond.h"
+#include "common/Thread.h"
+#include "common/Timer.h"
+#include "types.h"
+
+namespace rbd {
+namespace mirror {
+
+/**
+ * Manage deleted images from remote cluster.
+ */
+class ImageDeleter {
+public:
+  ImageDeleter(std::string remote_cluster_name, RadosRef local_cluster);
+  ~ImageDeleter();
+  ImageDeleter(const ImageDeleter&) = delete;
+  ImageDeleter& operator=(const ImageDeleter&) = delete;
+
+  void schedule_image_delete(uint64_t local_pool_id,
+                             const std::string& local_image_id,
+                             const std::string& global_image_id);
+
+private:
+
+  class ImageDeleterThread : public Thread {
+    ImageDeleter *m_image_deleter;
+  public:
+    ImageDeleterThread(ImageDeleter *image_deleter) :
+      m_image_deleter(image_deleter) {}
+    void *entry() {
+      m_image_deleter->run();
+      return 0;
+    }
+  };
+
+  class DeleteInfo {
+  public:
+    uint64_t local_pool_id;
+    std::string local_image_id;
+    std::string global_image_id;
+
+    DeleteInfo(uint64_t local_pool_id, const std::string& local_image_id,
+               const std::string& global_image_id) :
+      local_pool_id(local_pool_id), local_image_id(local_image_id),
+      global_image_id(global_image_id) {
+    }
+
+    std::string to_string();
+  };
+
+  enum FailedStep {
+    FAILED_ON_DISABLE_MIRRORING,
+    FAILED_ON_REMOVE_SNAPS,
+    FAILED_ON_REMOVE_IMAGE
+  };
+
+  class FailedDelete {
```

Comment:
Might be nice to combine `FailedDelete` and `DeleteInfo` into a single structure.  That way you could avoid having both the Thread and SafeTimer both running / deleting at the same time -- and you could re-use the existing rbd-mirror SafeTimer to just move failed items from a  "retry" queue back into the main queue (to be processed when the main thread becomes available).